akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala

```diff
@@ -237,20 +312,22 @@ private[akka] class MergeHub[T](perProducerBufferSize: Int)
           private[this] val id = idCounter.getAndIncrement()
 
           override def preStart(): Unit = {
-            if (!logic.isShuttingDown) {
-              logic.enqueue(Register(id, getAsyncCallback(onDemand)))
-
-              // At this point, we could be in the unfortunate situation that:
-              // - we missed the shutdown announcement and entered this arm of the if statement
-              // - *before* we enqueued our Register event, the Hub already finished looking at the queue
-              //   and is now dead, so we are never notified again.
-              // To safeguard against this, we MUST check the announcement again. This is enough:
-              // if the Hub is no longer looking at the queue, then it must be that isShuttingDown must be already true.
-              if (!logic.isShuttingDown) pullWithDemand()
-              else completeStage()
-            } else {
-              completeStage()
-            }
+            if (!logic.isDraining) {
+              if (!logic.isShuttingDown) {
+                logic.enqueue(Register(id, getAsyncCallback(onDemand)))
+
+                // At this point, we could be in the unfortunate situation that:
+                // - we missed the shutdown announcement and entered this arm of the if statement
+                // - *before* we enqueued our Register event, the Hub already finished looking at the queue
+                //   and is now dead, so we are never notified again.
+                // To safeguard against this, we MUST check the announcement again. This is enough:
+                // if the Hub is no longer looking at the queue, then it must be that isShuttingDown must be already true.
+                if (!logic.isShuttingDown) pullWithDemand()
+                else completeStage()
+              } else {
+                completeStage()
+              }
+            } else completeStage()
```

Comment:
make that a single `if (!isDraining && isShuttingDown)` instead?

Reply:
Agreed, I'll do that